mava/utils/sebulba.py

```diff
@@ -197,3 +203,357 @@ def __enter__(self) -> None:
+                min_size_to_sample * samples_per_insert + error_buffer
+              )
+            The offset is added so that the error tracked is for the insert/sample
+            ratio only takes into account operations occurring AFTER stage 1. If a
+            range (two float tuple) then the values are used without any offset.
+
+        Raises:
+          ValueError: If error_buffer is smaller than max(1.0, samples_per_inserts).
+        """
+        if isinstance(error_buffer, float) or isinstance(error_buffer, int):
+            offset = samples_per_insert * min_size_to_sample
+            min_diff = offset - error_buffer
+            max_diff = offset + error_buffer
+        else:
+            min_diff, max_diff = error_buffer
+
+        if samples_per_insert <= 0:
+            raise ValueError(f"samples_per_insert ({samples_per_insert}) must be > 0")
+
+        if max_diff - min_diff < 2 * max(1.0, samples_per_insert):
+            raise ValueError(
+                "The size of error_buffer must be >= max(1.0, samples_per_insert) as "
+                "smaller values could completely block samples and/or insert calls."
+            )
+
+        if max_diff < samples_per_insert * min_size_to_sample:
+            print(
+                f"{Fore.YELLOW}{Style.BRIGHT}The range covered by error_buffer is below "
+                "samples_per_insert * min_size_to_sample. If the sampler cannot "
+                "sample concurrently, this will result in a deadlock as soon as "
+                f"min_size_to_sample items have been inserted.{Style.RESET_ALL}"
+            )
+        if min_diff > samples_per_insert * min_size_to_sample:
+            raise ValueError(
+                "The range covered by error_buffer is above "
+                "samples_per_insert * min_size_to_sample. This will result in a "
+                "deadlock as soon as min_size_to_sample items have been inserted."
+            )
+
+        if min_size_to_sample < 1:
+            raise ValueError(
+                f"min_size_to_sample ({min_size_to_sample}) must be a positive integer"
+            )
+
+        super().__init__(
+            samples_per_insert=samples_per_insert,
+            min_size_to_sample=min_size_to_sample,
+            min_diff=min_diff,
+            max_diff=max_diff,
+        )
+
+
+# Modified from https://github.com/instadeepai/sebulba/blob/main/sebulba/core.py
+class OffPolicyPipeline(threading.Thread):
+    """
+    The `Pipeline` shards trajectories into learner devices,
+    ensuring trajectories are consumed in the right order to avoid being off-policy
+    and limit the max number of samples in device memory at one time to avoid OOM issues.
+    """
+
+    def __init__(
+        self,
+        config: DictConfig,
+        learner_sharding: Sharding,
+        key: jax.random.PRNGKey,
+        rate_limiter: RateLimiter,
+        init_transition: Transition,
+        lifetime: ThreadLifetime,
```

Comment:
While we're here, I think we should actually replace the lifetime with `threading.Event`, [it's a python builtin](https://docs.python.org/3/library/threading.html#event-objects) that does exactly what we want. Although maybe that should be a separate, I'll make an issue